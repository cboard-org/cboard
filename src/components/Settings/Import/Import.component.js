import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import Button from '@material-ui/core/Button';
import Paper from '@material-ui/core/Paper';
import List from '@material-ui/core/List';
import CircularProgress from '@material-ui/core/CircularProgress';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import ListItemSecondaryAction from '@material-ui/core/ListItemSecondaryAction';
import FullScreenDialog from '../../UI/FullScreenDialog';
import messages from './Import.messages';

import './Import.css';

const propTypes = {
  /**
   * Callback fired when clicking the import Cboard button
   */
  onImportClick: PropTypes.func.isRequired,
  /**
   * Callback fired when clicking the back button
   */
  onClose: PropTypes.func
};

class Import extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      loading: false
    };
  }

  onImportClick(event) {
    const doneCallback = () => {
      this.setState({ loading: false });
    };

    // https://reactjs.org/docs/events.html#event-pooling
    event.persist();

    this.setState({ loading: true, exportMenu: null }, () => {
      this.props.onImportClick(event, doneCallback);
    });
  }

  render() {
    const { onClose } = this.props;
    return (
      <div className="Import">
        <FullScreenDialog
          open
          title={<FormattedMessage {...messages.import} />}
          onClose={onClose}
        >
          <Paper>
            <List>
              <ListItem>
                <ListItemText
                  primary={<FormattedMessage {...messages.import} />}
                  secondary={
                    <FormattedMessage
                      {...messages.importSecondary}
                      values={{
                        link: (
                          <a href="http://www.openboardformat.org/">
                            OpenBoard
                          </a>
                        )
                      }}
                    />
                  }
                />
                <ListItemSecondaryAction>
                  <div className="Import__ButtonContainer">
                    {this.state.loading && (
                      <CircularProgress
                        size={25}
                        className="Import__ButtonContainer--spinner"
                        thickness={7}
                      />
                    )}
                    <Button
                      id="import-button"
                      component="span"
                      disabled={this.state.loading}
                    >
                      <label htmlFor="file">
                        <FormattedMessage {...messages.restore} />
                      </label>
                      <input
                        accept=".json,.obz,.obf,text/json,application/json"
                        id="file"
                        type="file"
                        style={{ display: 'none' }}
                        onChange={e => this.onImportClick(e)}
                      />
                    </Button>
                  </div>
                </ListItemSecondaryAction>
              </ListItem>
            </List>
          </Paper>
        </FullScreenDialog>
      </div>
    );
  }
}

Import.propTypes = propTypes;

export default Import;
