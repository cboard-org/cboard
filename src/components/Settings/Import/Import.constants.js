import {
  cboardImportAdapter,
  obzImportAdapter,
  obfImportAdapter
} from './Import.helpers';
import {
  CBOARD_EXT_PROPERTIES as EXPORT_CBOARD_EXT_PROPERTIES,
  CBOARD_EXT_PREFIX as EXPORT_CBOARD_EXT_PREFIX
} from '../Export/Export.constants';

export const IMPORT_CONFIG_BY_EXTENSION = {
  json: cboardImportAdapter,
  zip: obzImportAdapter,
  obz: obzImportAdapter,
  obf: obfImportAdapter
};

export const IMPORT_PATHS = {
  boards: 'boards/',
  images: 'images/'
};

export const CBOARD_EXT_PREFIX = EXPORT_CBOARD_EXT_PREFIX;

export const CBOARD_EXT_PROPERTIES = EXPORT_CBOARD_EXT_PROPERTIES;

export default {
  IMPORT_CONFIG_BY_EXTENSION,
  IMPORT_PATHS,
  CBOARD_EXT_PREFIX,
  CBOARD_EXT_PROPERTIES
};
