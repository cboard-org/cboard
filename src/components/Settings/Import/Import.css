.Import__ButtonContainer {
  position: relative;
}

.Import__ButtonContainer--spinner {
  vertical-align: middle;
}
