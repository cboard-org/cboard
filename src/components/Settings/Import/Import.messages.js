import { defineMessages } from 'react-intl';

export default defineMessages({
  import: {
    id: 'cboard.components.Settings.Import.import',
    defaultMessage: 'Import'
  },
  restore: {
    id: 'cboard.components.Settings.Import.restore',
    defaultMessage: 'Restore'
  },
  importSecondary: {
    id: 'cboard.components.Settings.Import.importSecondary',
    defaultMessage: 'Backup your boards'
  }
});
