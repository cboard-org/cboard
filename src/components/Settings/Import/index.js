export { default } from './Import.container';
