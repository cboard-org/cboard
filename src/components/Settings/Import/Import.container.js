import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { injectIntl, intlShape } from 'react-intl';
import { addBoards, changeBoard } from '../../Board/Board.actions';
import {
  upsertCommunicator,
  changeCommunicator
} from '../../Communicator/Communicator.actions';
import { switchBoard } from '../../Board/Board.actions';
import { showNotification } from '../../Notifications/Notifications.actions';
import Import from './Import.component';
import API from '../../../api';

export class ImportContainer extends PureComponent {
  static propTypes = {
    boards: PropTypes.array.isRequired,
    history: PropTypes.object.isRequired,
    intl: intlShape.isRequired
  };

  async updateLoadBoardsIds(boards, shouldUpdate = false) {
    const updatedBoards = await Promise.all(
      boards.map(async board => {
        const boardsToBeLoaded = {};
        const tilesToBeReplaced = {};
        const tempBoards = this.props.boards
          .concat(boards)
          .filter(b => b.prevId);
        let updated = false;
        board.tiles.forEach((tile, i) => {
          if (tile.loadBoard) {
            const boardToBeLoaded = tempBoards.find(
              tb => tb.prevId && tb.prevId === tile.loadBoard
            );
            if (boardToBeLoaded) {
              tilesToBeReplaced[i] = tile;
              boardsToBeLoaded[boardToBeLoaded.prevId] = boardToBeLoaded.id;
            }
          }
        });

        const tilesIndexes = Object.keys(tilesToBeReplaced);
        tilesIndexes.forEach(i => {
          const tile = board.tiles[i];

          if (boardsToBeLoaded[tile.loadBoard]) {
            board.tiles[i].loadBoard = boardsToBeLoaded[tile.loadBoard];

            updated = true;
          }
        });

        let boardToBeUpdated = board;
        if (shouldUpdate && updated) {
          boardToBeUpdated = await API.updateBoard(boardToBeUpdated);
        }

        return boardToBeUpdated;
      })
    );

    return updatedBoards;
  }

  async syncBoardsWithAPI(boards) {
    const { userData } = this.props;

    let boardsResponse = boards;
    if (userData.email) {
      const { email, name: author } = userData;

      boardsResponse = await Promise.all(
        boards.map(async board => {
          const boardToCreate = {
            ...board,
            email,
            author,
            isPublic: false,
            locale: this.props.intl.locale
          };

          const response = await API.createBoard(boardToCreate);

          if (board.id) {
            response.prevId = board.id;
          }

          return response;
        })
      );
    } else {
      const uuidv4 = await import('uuid/v4');
      boardsResponse.forEach(board => {
        if (board.id) {
          board.prevId = board.id;
        }

        board.id = uuidv4.default();
      });
    }

    boardsResponse = await this.updateLoadBoardsIds(
      boardsResponse,
      userData.email
    );

    this.props.addBoards(boardsResponse);
    await this.addBoardsToCommunicator(boardsResponse);
    this.props.switchBoard(boardsResponse[0].id);
  }

  async addBoardsToCommunicator(boards) {
    const { userData, currentCommunicator } = this.props;

    const communicatorBoards = new Set(
      currentCommunicator.boards.concat(boards.map(b => b.id))
    );
    let communicatorModified = {
      ...currentCommunicator,
      boards: Array.from(communicatorBoards)
    };

    if (userData && userData.authToken && userData.authToken.length) {
      communicatorModified = await API.updateCommunicator(communicatorModified);
    }

    this.props.upsertCommunicator(communicatorModified);
    this.props.changeCommunicator(communicatorModified.id);
  }

  async handleImportClick(e, doneCallback) {
    const { showNotification } = this.props;

    // Check for the various File API support.
    if (window.File && window.FileReader && window.FileList && window.Blob) {
      if (e.target.files.length > 0) {
        const file = e.target.files[0];
        const ext = file.name.match(/\.([^.]+)$/)[1];
        const importConstants = await import('./Import.constants');

        const importCallback =
          importConstants.IMPORT_CONFIG_BY_EXTENSION[ext.toLowerCase()];
        if (importCallback) {
          // TODO. Json format validation
          try {
            const jsonFile = await importCallback(file, this.props.intl);
            if (jsonFile.length) {
              const importHelpers = await import('./Import.helpers');
              await importHelpers.requestQuota(jsonFile);
              await this.syncBoardsWithAPI(jsonFile);
              showNotification('Backup restored successfuly.');
            }
          } catch (e) {
            console.error(e);
          }
        } else {
          alert('Please, select a valid file: json, obz, obf');
        }
      } else {
        console.warn('There is no selected file.');
      }
    } else {
      console.warn('The File APIs are not fully supported in this browser.');
    }

    if (doneCallback) {
      doneCallback();
    }
  }

  render() {
    const { boards, history } = this.props;

    return (
      <Import
        boards={boards}
        onImportClick={this.handleImportClick.bind(this)}
        onClose={history.goBack}
      />
    );
  }
}

const mapStateToProps = ({ board, communicator, app }) => {
  const activeCommunicatorId = communicator.activeCommunicatorId;
  const currentCommunicator = communicator.communicators.find(
    communicator => communicator.id === activeCommunicatorId
  );

  const { userData } = app;

  return {
    boards: board.boards,
    currentCommunicator,
    communicators: communicator.communicators,
    userData
  };
};

const mapDispatchToProps = {
  addBoards,
  changeBoard,
  switchBoard,
  upsertCommunicator,
  changeCommunicator,
  showNotification
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(injectIntl(ImportContainer));
