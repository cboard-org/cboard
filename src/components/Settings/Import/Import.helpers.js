import JSZipUtils from 'jszip-utils';
import JSZip from 'jszip';
import shortid from 'shortid';
import { IMPORT_PATHS, CBOARD_EXT_PREFIX } from './Import.constants';
import API from '../../../api';

function toCamelCase(scString = '') {
  const find = /(_\w)/g;
  const convertFn = matches => matches[1].toUpperCase();

  return scString.replace(find, convertFn);
}

async function readZip(file) {
  const zipBlob = await new Promise(resolve => {
    const reader = new FileReader();
    reader.onload = event => {
      if (event.target.readyState === 2) {
        try {
          resolve(new Blob([reader.result]));
        } catch (err) {
          resolve(null);
        }
      }
    };
    reader.readAsArrayBuffer(file);
  });

  if (zipBlob === null) {
    return Promise.reject();
  }

  const filePath = URL.createObjectURL(zipBlob);
  return new Promise(resolve => {
    JSZipUtils.getBinaryContent(filePath, (err, data) => {
      if (err) {
        resolve(err);
      } else {
        resolve(JSZip.loadAsync(data));
      }
    });
  });
}

function obfButtonToCboardButton(button) {
  const cboardButton = {
    id: button.id || shortid.generate(),
    label: button.label || ''
  };

  if (button['background_color']) {
    cboardButton.backgroundColor = button['background_color'];
  }

  if (button['border_color']) {
    cboardButton.borderColor = button['border_color'];
  }

  if (button.vocalization) {
    cboardButton.vocalization = button.vocalization;
  }

  if (button.action) {
    cboardButton.action = button.action;
  }

  return cboardButton;
}

async function getTilesData(obfBoard, boards = {}, images = {}) {
  const tiles = await Promise.all(
    obfBoard.buttons.map(async button => {
      const tileButton = obfButtonToCboardButton(button);

      if (button['load_board']) {
        const loadBoardData = button['load_board'];
        if (loadBoardData.path && boards[loadBoardData.path]) {
          tileButton.loadBoard = boards[loadBoardData.path].id;
        }
      }

      if (button['image_id']) {
        let imageID = button['image_id'];
        let image = obfBoard.images.find(image => image.id === imageID);

        if (image) {
          let imageData = image.data || null;
          if (image['content_type'] && image.path && images[image.path]) {
            // Certain OBF files have an incorrect MIME type for SVG files, so
            // the resulting data URI cannot be rendered. We need to fix the
            // MIME type ourselves.
            const contentType =
              image['content_type'] === 'image/svg'
                ? 'image/svg+xml'
                : image['content_type'];
            imageData = `data:${contentType};base64,${images[image.path]}`;
          }
          if (image.url) {
            tileButton.image = image.url;
          }
          if (imageData) {
            let url = imageData;
            try {
              const apiURL = await API.uploadFromDataURL(url, imageID, true);
              url = apiURL || url;
            } catch (err) {
              console.log(err.message);
            } finally {
              tileButton.image = url;
            }
          }
        }
      }

      const extKeys = Object.keys(button).filter(k =>
        k.startsWith(CBOARD_EXT_PREFIX)
      );
      extKeys.forEach(key => {
        const tileKey = toCamelCase(key.slice(CBOARD_EXT_PREFIX.length));
        tileButton[tileKey] = button[key];
      });

      return tileButton;
    })
  );

  return tiles;
}

async function obfToCboard(obfBoard, boards = {}, images = {}, allBoards = []) {
  const allBoardsIds = getBoardsIds(allBoards);
  if (allBoardsIds.includes(obfBoard.id)) {
    return undefined;
  }
  let tiles = [];
  if (obfBoard.buttons) {
    tiles = await getTilesData(obfBoard, boards, images);
  }
  let board = {
    id: obfBoard.id || shortid.generate(),
    tiles
  };

  if (obfBoard.grid) {
    board = { ...board, isFixed: true, grid: obfBoard.grid };
  }

  const extKeys = Object.keys(obfBoard).filter(k =>
    k.startsWith(CBOARD_EXT_PREFIX)
  );
  extKeys.forEach(key => {
    const tileKey = toCamelCase(key.slice(CBOARD_EXT_PREFIX.length));
    board[tileKey] = obfBoard[key];
  });
  if (typeof obfBoard.name !== 'undefined') {
    board.name = obfBoard.name;
  } else {
    board.name = 'unknown name';
  }

  if (obfBoard.locale) {
    board.locale = obfBoard.locale;
  }

  return board;
}

function getBoardsIds(boards) {
  const allBoardsIds = [];
  boards.forEach(board => {
    if (typeof board.id !== 'undefined') {
      allBoardsIds.push(board.id);
    }
    if (typeof board.prevId !== 'undefined') {
      allBoardsIds.push(board.prevId);
    }
  });
  return allBoardsIds;
}

export async function cboardImportAdapter(file, intl, allBoards) {
  const reader = new FileReader();
  return new Promise((resolve, reject) => {
    reader.onload = async event => {
      if (event.target.readyState === 2) {
        try {
          const boards = JSON.parse(reader.result);
          const allBoardsIds = getBoardsIds(allBoards);
          const fboards = boards.filter(
            board =>
              (typeof board.ext_cboard_hidden === 'undefined' ||
                !board.ext_cboard_hidden) &&
              board.id !== 'root' &&
              !allBoardsIds.includes(board.id)
          );
          resolve(fboards);
        } catch (err) {
          reject(err);
        }
      }
    };
    reader.readAsText(file);
  });
}

export async function obzImportAdapter(file, intl, allBoards) {
  const zipFile = await readZip(file);
  if (typeof zipFile !== 'object') {
    throw new Error(zipFile);
  }

  const keys = Object.keys(zipFile.files);
  const boardKeys = keys.filter(
    k => !zipFile.files[k].dir && k.endsWith(IMPORT_PATHS.boards)
  );
  const imageKeys = keys.filter(
    k => !zipFile.files[k].dir && k.startsWith(IMPORT_PATHS.images)
  );
  const boards = {};
  const images = {};
  const allBoardsIds = getBoardsIds(allBoards);
  await Promise.all(
    keys.map(async k => {
      const isBoard = boardKeys.indexOf(k) >= 0;

      if (!isBoard && imageKeys.indexOf(k) < 0) {
        return Promise.resolve();
      }

      const type = isBoard ? 'text' : 'base64';
      let result = null;
      try {
        result = await zipFile.files[k].async(type);

        if (isBoard) {
          const tempBoard = JSON.parse(result);
          if (
            (typeof tempBoard.ext_cboard_hidden === 'undefined' ||
              !tempBoard.ext_cboard_hidden) &&
            tempBoard.id !== 'root' &&
            !allBoardsIds.includes(tempBoard.id)
          ) {
            boards[k] = tempBoard;
          }
        } else {
          if (k.startsWith('images//')) {
            images[k.substring(7)] = result;
          } else {
            images[k] = result;
          }
        }
      } catch (e) {}

      return result;
    })
  );

  const cboardBoards = [];
  for (let key in boards) {
    const board = await obfToCboard(boards[key], boards, images);
    cboardBoards.push(board);
  }

  return cboardBoards;
}

export async function obfImportAdapter(file, intl, allBoards) {
  const reader = new FileReader();
  const jsonFile = await new Promise(resolve => {
    reader.onload = event => {
      if (event.target.readyState === 2) {
        try {
          const jsonFile = JSON.parse(reader.result);
          resolve(jsonFile);
        } catch (err) {
          resolve(err);
        }
      }
    };
    reader.readAsText(file);
  });

  if (typeof jsonFile !== 'object') {
    throw new Error(jsonFile);
  }

  const board = await obfToCboard(jsonFile, {}, {}, allBoards);
  if (board) {
    return [board];
  }
  return [];
}

export async function requestQuota(json) {
  const size = JSON.stringify(json).length;
  if (size > 1024 * 1024 * 4) {
    const requestQuotaAvailable =
      navigator &&
      navigator.webkitPersistentStorage &&
      navigator.webkitPersistentStorage.requestQuota;
    if (requestQuotaAvailable) {
      try {
        await new Promise((resolve, reject) => {
          navigator.webkitPersistentStorage.requestQuota(
            size * 2,
            grantedSize => {
              if (grantedSize >= size) {
                resolve();
              } else {
                reject(`Granted size is below the limit: ${grantedSize}`);
              }
            },
            err => reject(`Request quota error: ${err}`)
          );
        });
      } catch (e) {
        throw new Error(e);
      }
    } else {
      throw new Error("Can't request quota");
    }
  }

  return size;
}
