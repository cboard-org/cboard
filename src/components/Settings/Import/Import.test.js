import React from 'react';
import { shallowMatchSnapshot } from '../../../common/test_utils';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';

import Import from './Import.component';

jest.mock('./Import.messages', () => {
  return {
    import: {
      id: 'cboard.components.Settings.Import.import',
      defaultMessage: 'Import'
    },
    restore: {
      id: 'cboard.components.Settings.Import.restore',
      defaultMessage: 'Restore'
    },
    exportSecondary: {
      id: 'cboard.components.Settings.Import.importSecondary',
      defaultMessage: 'Backup your boards'
    }
  };
});

const COMPONENT_PROPS = {
  onImportClick: () => {},
  onClose: () => {}
};

describe('Import tests', () => {
  test('default renderer', () => {
    shallowMatchSnapshot(<Import {...COMPONENT_PROPS} />);
  });

  test('loading behavior', () => {
    const wrapper = shallow(<Import {...COMPONENT_PROPS} />);
    let tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();

    wrapper.instance().onImportClick = jest.fn(type => {
      wrapper.setState({ loading: true });
    });

    let spinnerWrapper = wrapper.find('.Import__ButtonContainer--spinner');
    expect(spinnerWrapper.length).toBe(0);

    const importButton = wrapper.find('#import-button input');
    importButton.simulate('change', { currentTarget: 'someElement' });

    spinnerWrapper = wrapper.find('.Import__ButtonContainer--spinner');
    expect(spinnerWrapper.length).toBe(1);

    expect(wrapper.find('#import-button').get(0).props.disabled).toBe(true);

    tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
  });
});
