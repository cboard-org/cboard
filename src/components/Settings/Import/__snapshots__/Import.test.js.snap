// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Import tests default renderer 1`] = `
<div
  className="Import"
>
  <WithStyles(FullScreenDialog)
    onClose={[Function]}
    open={true}
    title={
      <FormattedMessage
        defaultMessage="Import"
        id="cboard.components.Settings.Import.import"
        values={Object {}}
      />
    }
  >
    <WithStyles(Paper)>
      <WithStyles(List)>
        <WithStyles(ListItem)>
          <WithStyles(ListItemText)
            primary={
              <FormattedMessage
                defaultMessage="Import"
                id="cboard.components.Settings.Import.import"
                values={Object {}}
              />
            }
            secondary={
              <FormattedMessage
                values={
                  Object {
                    "link": <a
                      href="http://www.openboardformat.org/"
                    >
                      OpenBoard
                    </a>,
                  }
                }
              />
            }
          />
          <WithStyles(ListItemSecondaryAction)>
            <div
              className="Import__ButtonContainer"
            >
              <WithStyles(Button)
                component="span"
                disabled={false}
                id="import-button"
              >
                <label
                  htmlFor="file"
                >
                  <FormattedMessage
                    defaultMessage="Restore"
                    id="cboard.components.Settings.Import.restore"
                    values={Object {}}
                  />
                </label>
                <input
                  accept=".json,.obz,.obf,text/json,application/json"
                  id="file"
                  onChange={[Function]}
                  style={
                    Object {
                      "display": "none",
                    }
                  }
                  type="file"
                />
              </WithStyles(Button)>
            </div>
          </WithStyles(ListItemSecondaryAction)>
        </WithStyles(ListItem)>
      </WithStyles(List)>
    </WithStyles(Paper)>
  </WithStyles(FullScreenDialog)>
</div>
`;

exports[`Import tests loading behavior 1`] = `
<div
  className="Import"
>
  <WithStyles(FullScreenDialog)
    onClose={[Function]}
    open={true}
    title={
      <FormattedMessage
        defaultMessage="Import"
        id="cboard.components.Settings.Import.import"
        values={Object {}}
      />
    }
  >
    <WithStyles(Paper)>
      <WithStyles(List)>
        <WithStyles(ListItem)>
          <WithStyles(ListItemText)
            primary={
              <FormattedMessage
                defaultMessage="Import"
                id="cboard.components.Settings.Import.import"
                values={Object {}}
              />
            }
            secondary={
              <FormattedMessage
                values={
                  Object {
                    "link": <a
                      href="http://www.openboardformat.org/"
                    >
                      OpenBoard
                    </a>,
                  }
                }
              />
            }
          />
          <WithStyles(ListItemSecondaryAction)>
            <div
              className="Import__ButtonContainer"
            >
              <WithStyles(Button)
                component="span"
                disabled={false}
                id="import-button"
              >
                <label
                  htmlFor="file"
                >
                  <FormattedMessage
                    defaultMessage="Restore"
                    id="cboard.components.Settings.Import.restore"
                    values={Object {}}
                  />
                </label>
                <input
                  accept=".json,.obz,.obf,text/json,application/json"
                  id="file"
                  onChange={[Function]}
                  style={
                    Object {
                      "display": "none",
                    }
                  }
                  type="file"
                />
              </WithStyles(Button)>
            </div>
          </WithStyles(ListItemSecondaryAction)>
        </WithStyles(ListItem)>
      </WithStyles(List)>
    </WithStyles(Paper)>
  </WithStyles(FullScreenDialog)>
</div>
`;

exports[`Import tests loading behavior 2`] = `
<div
  className="Import"
>
  <WithStyles(FullScreenDialog)
    onClose={[Function]}
    open={true}
    title={
      <FormattedMessage
        defaultMessage="Import"
        id="cboard.components.Settings.Import.import"
        values={Object {}}
      />
    }
  >
    <WithStyles(Paper)>
      <WithStyles(List)>
        <WithStyles(ListItem)>
          <WithStyles(ListItemText)
            primary={
              <FormattedMessage
                defaultMessage="Import"
                id="cboard.components.Settings.Import.import"
                values={Object {}}
              />
            }
            secondary={
              <FormattedMessage
                values={
                  Object {
                    "link": <a
                      href="http://www.openboardformat.org/"
                    >
                      OpenBoard
                    </a>,
                  }
                }
              />
            }
          />
          <WithStyles(ListItemSecondaryAction)>
            <div
              className="Import__ButtonContainer"
            >
              <WithStyles(CircularProgress)
                className="Import__ButtonContainer--spinner"
                size={25}
                thickness={7}
              />
              <WithStyles(Button)
                component="span"
                disabled={true}
                id="import-button"
              >
                <label
                  htmlFor="file"
                >
                  <FormattedMessage
                    defaultMessage="Restore"
                    id="cboard.components.Settings.Import.restore"
                    values={Object {}}
                  />
                </label>
                <input
                  accept=".json,.obz,.obf,text/json,application/json"
                  id="file"
                  onChange={[Function]}
                  style={
                    Object {
                      "display": "none",
                    }
                  }
                  type="file"
                />
              </WithStyles(Button)>
            </div>
          </WithStyles(ListItemSecondaryAction)>
        </WithStyles(ListItem)>
      </WithStyles(List)>
    </WithStyles(Paper)>
  </WithStyles(FullScreenDialog)>
</div>
`;
